blog/content/post/2021/03/23/foo.md

```diff
@@ -0,0 +1,90 @@
+---
+title: "Running Choria Accross the Globe"
+title: "Fighting Latency with Choria"
+date: 2021-03-23T00:00:00+01:00
+tags: ["leafnodes"]
+draft: false
+author: Romain Tartière
+---
+
+Because Choria allows you to manage nodes spread all around the world, and because you might be working from your laptop, far away from the (bad) Wi-Fi access points that connects you through (bad) PLC to the (bad) internet connexion from the (not bad) island you are on, you may experience inconvenient lattency and unreliabilities.
+
+The reason is quite simple: while the Choria servers maintain a permanent connexion with the message broker, the Choria client has to establish a new connexion with the middleware for each request.  Lattency and packet loss do not help with establishing TLS encrypted connexions in a timely fashion.
+
+But _good news everyone_!  NATS — the messaging system Choria is built on — has built-in support for so-called _leaf nodes_ which offer a solution to this problem.
+
+<!--more-->
+
+A _leaf node_ acts as a broker node.  It accepts client conexions on port 4222 and allow messages to be distributed to the server nodes.  The key point is that it maintains a permanent connexion with the Choria borker, so when a client connect to it, it is immediatly able to send messages through the middleware.
+
+As a consequence, installing a leaf node nearby a Choria client (on the LAN in the office if multiple employees need to access the message broker, or maybe directly on your laptop if you happen to move around) allows the Choria clients to establish the expensive TLS handshake with a node next to them rather than one far far away, making the whole process faster and more reliable.
+
+FIXME: Add a drawing here to compare both setup.  Highlight permanent vs. ponctual connexions and also indicate ports used
+
+Let's see how to proceed to deploy a leaf node!
+
+## Deploying a leaf node
+
+FIXME: Tell which version introduce this change
+
+It is now possible to use the `choria-choria` puppet module to setup leaf nodes, this section shows how to proceed.
+
+Before starting, you might want to measure the lattency when you start to compare with what you have at the end.  Let's use `choria ping` for this purpose:
+
+```
+romain@zappy % choria ping
+[...]
+---- ping statistics ----
+47 replies max: 378.357 min: 172.005 avg: 304.544 overhead: 1940.673
+```
+
+All numbers are expressed in milliseconds.
+
+The _max_ / _min_ / _avg_ figures are not the one we are interested in here because they measure the lattency between the choria broker and the choria servers.  The _overhead_ is what we want to improve, nearly 2 seconds in my case.
```

Comment:
latency between choria client and choria servers via the broker once already connected

Reply:
Not 100% sure if it is accurate, I added this clarification which would need proofreading:

> The _overhead_ is what we want to improve: _the time needed by the client to establish a connexion to the broker node_, nearly 2 seconds in my case.

Reply:
Yes close enough - the other thing it includes is the security layer time, in my case my AAA server is in the netherlands and that round trip is also included in overhead.  You runnin tradditional cert based access dont have much measuarable time there (micro seconds)

Reply:
so for all intents your description is correct :)